lib/shipit/task_commands.rb

```diff
@@ -28,22 +28,20 @@ def steps
     end
 
     def env
-      super.merge(
-        'ENVIRONMENT' => @stack.environment,
-        'BRANCH' => @stack.branch,
-        'SHIPIT_USER' => "#{@task.author.login} (#{normalized_author_name}) via Shipit",
-        'EMAIL' => @task.author.email,
-        'BUNDLE_PATH' => Rails.root.join('data', 'bundler').to_s,
-        'SHIPIT_LINK' => @task.permalink,
-        'LAST_DEPLOYED_SHA' => @stack.last_deployed_commit.sha,
-        'TASK_ID' => @task.id.to_s,
-        'DEPLOY_URL' => @stack.deploy_url,
-        'IGNORED_SAFETIES' => @task.ignored_safeties? ? '1' : '0',
-        'GIT_COMMITTER_NAME' => @task.user&.name || Shipit.committer_name,
-        'GIT_COMMITTER_EMAIL' => @task.user&.email || Shipit.committer_email,
-        'GITHUB_REPO_OWNER' => @stack.repository.owner,
-        'GITHUB_REPO_NAME' => @stack.repository.name,
-      ).merge(deploy_spec.machine_env).merge(@task.env)
+      super
+        .merge(
+          'SHIPIT_USER' => "#{@task.author.login} (#{normalized_author_name}) via Shipit",
+          'EMAIL' => @task.author.email,
+          'BUNDLE_PATH' => Rails.root.join('data', 'bundler').to_s,
+          'SHIPIT_LINK' => @task.permalink,
+          'TASK_ID' => @task.id.to_s,
+          'IGNORED_SAFETIES' => @task.ignored_safeties? ? '1' : '0',
+          'GIT_COMMITTER_NAME' => @task.user&.name || Shipit.committer_name,
+          'GIT_COMMITTER_EMAIL' => @task.user&.email || Shipit.committer_email,
+        )
+        .merge(deploy_spec.machine_env)
+        .merge(@stack.env)
+        .merge(@task.env)
```

Comment:
My last worry with the PR is here.

Environment variables precedence is important.

I don't think `Stack#env` should be able to override `Task#env`.

Intuitively I'd say the order should be:

  - `Commands#env` (`super` in this context, mostly comes from the global config file)
  - `Stack#env`
  - `TaskCommands#env`
  - `DeploySpec#machine_env`
  - `Task#env`

But I'm open to diverging opinions.

Reply:
Entirely reasonable. Thanks for the feedback! Something like https://github.com/Shopify/shipit-engine/pull/1018/commits/09cc2991332935537453101c112f6855faf72a84 instead, then?

Reply:
Yup 👍

Reply:
Many, many thanks, friend. 🍻